src/model/Model.js

```diff
@@ -0,0 +1,347 @@
+   */
+  static setup(faunaClassName, fields={}) {
+    this.faunaClassName = faunaClassName
+    /**
+     * {@link Ref} for the class itself.
+     *
+     * `instance.ref` should be the same as `new Ref(instance.constructor.classRef, instance.id)`.
+     */
+    this.classRef = new Ref('classes', faunaClassName)
+    /** Object of all fields assigned to this class. */
+    this.fields = {}
+    for (const fieldName in fields)
+      this.addField(fieldName, fields[fieldName])
+  }
+
+  /**
+   * Adds a new field to the class, making getters and setters.
+   *
+   * @param {string} fieldName
+   *   Name for the field. A getter and setter will be made with this name.
+   *   If `fieldOpts.path` is not defined, it defaults to `['data', fieldName]`.
+   * @param {object} fieldOpts
+   *   Parameters for the {@link Field} constructor.
+   */
+  static addField(fieldName, fieldOpts={}) {
+    if (fieldName === 'ref' || fieldName === 'ts')
+      throw new Error('Forbidden field name.')
+
+    if (fieldOpts.path == null)
+      fieldOpts.path = ['data', fieldName]
+    const field = new Field(fieldOpts)
+    this.fields[fieldName] = field
+
+    const {get, set} = field.codec === null ?
+      {
+        get() {
+          return getPath(field.path, this._current)
+        },
+        set(value) {
+          setPath(field.path, value, this._current)
+        }
+      } : {
+        get() {
+          if (fieldName in this._cache)
+            return this._cache[fieldName]
+          else {
+            const encoded = getPath(field.path, this._current)
+            const decoded = field.codec.decode(encoded, this)
+            this._cache[fieldName] = decoded
+            return decoded
+          }
+        },
+        set(value) {
+          // Clear any previous cached decoded value.
+          delete this._cache[fieldName]
+          // Codecs are not guaranteed to have decode(encode(value)) === value.
+          // So can't set _cache until getter is called.
+          const encoded = field.codec.encode(value, this)
+          setPath(field.path, encoded, this._current)
+        }
+      }
+    Object.defineProperty(this.prototype, fieldName, {get, set})
+  }
+
+  /**
+   * Initialize (but do not save) a new instance.
+   * @param {Client} client
+   * @param {object} data Fields to be set.
+   */
+  constructor(client, data) {
+    /** Client instance that the model uses to save to the database. */
+    this.client = client
+
+    this._original = {}
+    this._initState()
+
+    for (const fieldName in data) {
+      if (!(fieldName in this.constructor.fields))
+        throw new InvalidValue(`No such field ${fieldName}`)
+      // This calls the field's setter.
+      this[fieldName] = data[fieldName]
+    }
+  }
+
+  /** {@link Ref} of this instance in the database. Fails if {@link isNewInstance}. */
+  get ref() {
+    if (this.isNewInstance())
+      throw new InvalidQuery('Instance has not been saved to the database, so no ref exists.')
```

Comment:
I feel like getters should return null, rather than throw, in this case? I kind of feel like its a bit weird for getters to throw, but maybe its not that weird.